syntax = "proto3";

package eigr.actors;

import "google/protobuf/any.proto";

option java_package = "io.eigr.functions.protocol.actors";
option go_package = "github.com/eigr/go-support/eigr/actors;actors";

message Registry {
    map<string, Actor> actors = 1;
}

message ActorSystem {
    string name = 1;
    Registry registry = 2;
}

// A strategy for save state.
message ActorSnapshotStrategy {
    oneof strategy {
        // the timeout strategy.
        TimeoutStrategy timeout = 1;
    }
}

// A strategy which a user function's entity is passivated.
message ActorDeactivateStrategy {
    oneof strategy {
        // the timeout strategy.
        TimeoutStrategy timeout = 1;
    }
}

// A strategy based on a timeout.
message TimeoutStrategy {
    // The timeout in millis
    int64 timeout = 1;
}

message ActorState {
    map<string, string> tags = 1;
    google.protobuf.Any state = 2;
}

message Actor {
    string name = 1;
    bool persistent = 2;
    ActorState state = 3;
    ActorSnapshotStrategy snapshot_strategy = 4;
    ActorDeactivateStrategy deactivate_strategy = 5;
}